### PR TITLE
Transition auth state to MobX

### LINKED ISSUE #247: Move auth data model to MobX
**What needs to be done? Why does it need to be done?**
We are strongly considering [MobX](https://mobx.js.org/README.html) as our company-wide standard package for state management and data modeling in our frontend applications. As a proof-of-concept and introduction to this new library, transition the existing authentication data model to a MobX data store. 

This feature is already implemented and works fine but as the only bit of state that currently exists in Spotlight v1, it's a good candidate for this because it is small and clear in scope, well-understood, and implements patterns that are common across multiple of our web applications. The hope is that this will be a good introduction to the library in advance of building out more substantial data models for content and metrics, and also that it will prevent us from having a legacy data model existing alongside MobX models

There are a few areas of particular interest in this implementation: 
- adding MobX to our frontend stack and generally assessing its fit
- creating a strong separation between state and UI, as discussed in various community examples of how best to use MobX. In particular, implement a flow where the UI can be fully derived from the MobX store and does not interact directly with the page URL or even the Auth0 client libraries
- achieve feature parity with no regression from the existing auth solution
- give at least some consideration to how this auth solution could generalize to pulse-dashboard, without necessarily solving for that use case

### PR DESCRIPTION
**Pull request by macfarlandian**
## Description of the change

Updates the authentication feature to use a MobX data store rather than React state, as a proof-of-concept and introduction to integrating MobX into our web application stack. 

Overall I am quite happy with this change architecturally, even given its limited scope. To wit: 

- I took inspiration from the design patterns in [this article by a MobX maintainer](https://medium.com/hackernoon/how-to-decouple-state-and-ui-a-k-a-you-dont-need-componentwillmount-cc90b787aa37#.7zvpxt746) to have the URLs drive the model and the model drive the UI. While the URL piece is of limited relevance to this feature, I find the overall separation to be quite clean. 
- The implementation of `AuthWall` (the component consuming this model) is notably free of the convoluted context management that the previous implementation required. In addition, the dependency on `auth0-react` has been removed, and the React application itself is now blissfully unaware that Auth0 even exists; all it consumes are a few boolean flags concerning authorization state and an `authorize` method to call as needed.
- I tried to think a bit about what would help the `UserStore` model be portable across our Auth0 tenants. The most concrete step that followed from that was to keep it uncoupled from the environment variables that we derive the Auth0 settings from, in favor of requiring those settings in its `constructor`. The piece of the application that mounts the data stores is responsible for providing those arguments. This also made the model pretty straightforward to test in isolation with class instances.
- The React tests are much more holistic now. Other than continuing to mock the Auth0 client library to prevent it from making network requests and doing other things that JSDOM does not like, it stays away from implementation details; we set realistic environment variables and inspect the rendered output in a way that parallels the real browser experience much more closely. There is some test overlap in terms of the authentication behavior involved, but thanks to the clean separation of view and model, the expectations themselves are not really redundant.
- I did have to update some of the testing dependencies here to resolve some JSDOM issues and get the latest async testing utilities. I also had to update the CI config to make sure we were using a supported node version (the `setup-node` action was pinned to a version that actually ignored the `node-version` argument, whoops!)

Anyway I am coming away from this feeling pretty good about MobX, but I am very curious to see what you all think!

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #247 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: spotlight-client/src/stores/UserStore.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+  }
+
+  /**
+   * If user already has a valid Auth0 credential, this method will retrieve it
+   * and update class properties accordingly. If not, user will be redirected
+   * to the Auth0 login domain for fresh authentication.
+   * Returns an Error if Auth0 configuration is not present.
+   */
+  async authorize(): Promise<Error | void> {
```
Comment (jessex) [inline spotlight-client/src/stores/UserStore.ts:80]: Did you happen to see the [Auth0 post](https://auth0.com/blog/building-and-securing-mobx-apps-with-auth0/) about Auth0 x React x MobX? If not, how does your approach here compare to what they described therein? In particular, I want to make sure we are writing as little auth-related logic as possible ourselves. I want to delegate as much as possible either into the auth layer itself or into vetted code written by auth professionals.

**Review by jessex — COMMENTED** [review]
I didn't do the _most_ thorough review admittedly, but I did look through the new architecture and I'm a fan! I like the separation of concerns, I like the expanded test coverage. Just want to make sure we're doing only as much work as is appropriate for us to do ourselves.

File: spotlight-client/src/stores/UserStore.ts (hunk 1 above)
Comment (macfarlandian) [reply spotlight-client/src/stores/UserStore.ts:80]: I did use it as a reference to some degree, but I am going to read it again through this particular lens and report back

Comment (macfarlandian) [reply spotlight-client/src/stores/UserStore.ts:80]: OK so after another review there are a few things I would change:
- I think I should just call the `createAuth0Client` function instead of initializing it manually, so that it does the Officially Recommended Things (which I still find to be largely opaque and mysterious). whatever reason I had for not doing that originally does not seem to be applicable anymore
- I was a bit hesitant to create a dedicated callback url like they do in this example because I thought the application should be able to authorize you "in place". looking at it again this morning though, I am not so convinced that makes sense, since it requires us to expect that auth0 is going to add certain search params to the url. (I adapted that logic from the COVID site but inspecting that seems unnecessarily brittle in this context.) There should be other ways to make sure we return you to the original page after authorizing you, if in fact we even care to do that

The only other significant difference between this code and that example is that they have you click a button to log in whereas we just trigger it automatically. There is no difference there in the underlying interactions with the Auth0 SDK, it's just a user experience difference.

Thanks for calling this out! Definitely something that got a bit lost in the weeds

Comment (macfarlandian) [reply spotlight-client/src/stores/UserStore.ts:80]: ok well this turned out to be an unexpected journey .... I did adopt the `createAuth0Client` factory function, which is a definite improvement; in the absence of that the Auth0 SDK was doing some extra redirects to authenticate you which was making for a very choppy experience. 

Creating a callback URL will require us to figure out what the URL routing story will be, which turned out be not very straightforward; there doesn't seem to be any open-source solution out there that Just Works in the way I was looking for (I evaluated like three or four, all of which eventually fell over in Extremely Not Fun ways). I don't think there is anything really wrong with the existing solution for callback handling, since it is also an Auth0-recommended approach and one we have used elsewhere. This could be something to revisit later once we have URL routing set up, but since this only affects staging for now anyway I don't think it's worth spending any more time on right now or adding to the scope of this feature.

**Review by jessex — APPROVED** [review]
Excited to see this architectural innovation and I'm on board with giving this a shot with Auth0 in this app!

File: spotlight-client/src/stores/RootStore.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+export function isAuthEnabled(): boolean {
+  return process.env.REACT_APP_AUTH_ENABLED === "true";
+}
+
+export default class RootStore {
+  userStore: UserStore;
+
+  constructor() {
+    this.userStore = new UserStore({
+      authSettings: getAuthSettings(),
```
Comment (jovergaag) [inline spotlight-client/src/stores/RootStore.ts:48]: Why did you decide to pass in `authSettings` and `isAuthEnabled` here instead of keeping all of the auth code in the `UserStore`? For easier testing?

File: spotlight-client/src/stores/UserStore.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      ignoreQueryPrefix: true,
+    });
+    if (urlQuery.code && urlQuery.state) {
+      await auth0.handleRedirectCallback();
+      // auth0 params are single-use, must be removed or they'll cause errors
+      const newUrl = new URL(window.location.href);
+      delete urlQuery.code;
+      delete urlQuery.state;
+      newUrl.search = qs.stringify(urlQuery, { addQueryPrefix: true });
+      window.history.replaceState({}, document.title, newUrl.href);
```
Comment (jovergaag) [inline spotlight-client/src/stores/UserStore.ts:97]: I'm wondering if this is more complicated than it needs to be. In pulse-dashboard we use:
```
const { appState } = await auth0FromHook.handleRedirectCallback();
window.history.replaceState(
    {},
    document.title,
    appState && appState.targetUrl
      ? appState.targetUrl
      : window.location.pathname
);
```

Are there other requirements here that require the manual manipulation of the `urlQuery`?

**Review by jovergaag — APPROVED** [review]
This is super slick. I really like how clean the `AuthWall` and `UserStore` ended up. I think this is definitely the direction we want to be heading in overall. 

There's not much store manipulation in this PR due to the simplicity of the user store, I'm excited to see how the more involved MobX stores will turn out.

Just a couple of questions/comments. Nothing blocking here.

File: spotlight-client/src/stores/RootStore.ts (hunk 2 above)
Comment (macfarlandian) [reply spotlight-client/src/stores/RootStore.ts:48]: not primarily for that reason, though it did improve the tests as a bonus ... I didn't want to directly couple the UserStore class to the environment, I wanted it to stay as self-contained as possible. The RootStore, being at the top of the tree, seemed like a more appropriate place to be interacting directly with the environment to let its children remain more modular.

File: spotlight-client/src/stores/UserStore.ts (hunk 3 above)
Comment (macfarlandian) [reply spotlight-client/src/stores/UserStore.ts:97]: no, you are right ... the original thinking here is that we are likely to be using query params for some application state and I didn't want to clobber them. But then I got sidetracked and never got around to replacing this with something better!

**Review by jovergaag — APPROVED** [review]
👍

**Comment by macfarlandian** [comment]
sorry this commit history got a little hard to follow because I had a few PRs stacked on top of each other ... allow me to explain: 

We had decided to put this one on ice for a while while we leveled up our MobX understanding ... something didn't smell quite right. Now, one month wiser, I think I'm ready to put it to bed. Here's what I have changed: 

- merged the root store from this PR into the new root store from `master` that includes the Tenant data models
- with MobX linting activated, I was getting some warnings about how the `UserStore` observables were being accessed. To resolve those, the side effect in `AuthWall` is now reactive, as are the UserStore tests
- rather than returning a fetch error via the `authorize` Promise, the error is now an observable property of the `UserStore`. Now we can just observe that value instead of having to await the promise, which is less imperative and more reactive. as all the cool kids want to be 😎

File: spotlight-client/src/AuthWall/AuthWall.tsx [hunk 4]
```diff
@@ -15,81 +15,54 @@
...
- * in an Auth0Provider to enable the Auth0 React context.
- * If auth is disabled, renders its children unwrapped.
+ * Verifies authorization before rendering its children.
  */
-const AuthProvider: React.FC<{ enabled: boolean }> = ({
-  children,
-  enabled,
-}) => {
-  const authSettings = getAuthSettings();
+const AuthWall: React.FC = ({ children }) => {
```
Comment (jovergaag) [inline spotlight-client/src/AuthWall/AuthWall.tsx:31]: So clean and crisp. Very nice!

**Review by jovergaag — APPROVED** [review]
These changes look great. It's really coming together!

I tested both REACT_APP_AUTH_ENABLED=true and REACT_APP_AUTH_ENABLED=false, went through the whole workflow of adding/verifying user. All looks good!
